### PR TITLE
ACME Directory Resource

### PR DESCRIPTION
**Pull request by kuba**
Implements https://github.com/letsencrypt/acme-spec/blob/master/draft-barnes-acme.md#directory and serves as important basis for future ACME spec PRs.

This should be pulled in only after Boulder devs remove the following line: https://github.com/letsencrypt/boulder/blob/56ff42de06ee87265c8ec02d59e909a0f4b8873a/test/amqp-integration-test.py#L106. Otherwise it will cause integration test problems.

### CODE REVIEW

**Comment by kuba** [comment]
https://github.com/letsencrypt/boulder/pull/766 removes the aforementioned line

File: acme/acme/messages.py [hunk 1]
```diff
@@ -128,6 +127,56 @@ class Identifier(jose.JSONObjectWithFields):
...
+            # dictionary, whose keys are the "resource" values listed
+            # in {{https-requests}}'z
+            raise ValueError('Wrong directory fields')
+        # TODO: check that everything is an absolute URL; acme-spec is
+        # not clear on that
+        self._jobj = canon_jobj
+
+    def __getattr__(self, name):
+        try:
+            return self[name.replace('_', '-')]
```
Comment (jdkasten) [inline acme/acme/messages.py]: It is unclear to me why we are replacing _ with - 
I realize why it works, but it seems to add some deal of complication.  I know it has proliferated into other parts of the code (plugins)  

What is the exact internal naming convention? Can it be documented somewhere?

**Comment by jdkasten** [comment]
At the end of all of this... we should probably make UML diagrams to go with the ACME resources and messages.  I drew it all out once... (many ACME iterations ago)... and it really helps.

**Comment by jdkasten** [comment]
Thanks!

File: acme/acme/messages.py (hunk 1 above)
Comment (kuba) [inline acme/acme/messages.py]: `__getattr__` is a magic for `directory.new_reg` type of access. `directory.new-reg` is parsed by Python as `directory.new - reg`, i.e. subtraction expression. I should document it in the `Directory.__doc__`...

**Pull request merged**